src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs

```diff
@@ -181,7 +181,9 @@ public static Task<SyntaxTreeIndex> GetSyntaxTreeIndexAsync(this Document docume
             if (document.Project.TryGetCompilation(out var compilation))
             {
                 // We already have a compilation, so at this point it's fastest to just get a SemanticModel
-                return await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                GC.KeepAlive(compilation);
```

Comment:
So *technically* this isn't required. `GetSemanticModelAsync` will do the right thing in both cases (compilation is still available, or compilation was serialized to temporary storage and needs to be "reconstituted"). Given that this is not a correctness issue, I'd like to find a "pit of success" style solution if possible, where the best thing is also the intuitive thing. @jasonmalinowski ideas for that?

@pentp I sent you a message on Gitter too. 😄